// @ika.xyz/sdk/dist/esm/generated/ika_dwallet_2pc_mpc/ika_dwallet_2pc_mpc_init.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as _package from "./deps/sui/package.js";
const $moduleName = "@local-pkg/2pc-mpc::ika_dwallet_2pc_mpc_init";
const IKA_DWALLET_2PC_MPC_INIT = new MoveStruct({ name: `${$moduleName}::IKA_DWALLET_2PC_MPC_INIT`, fields: {
  dummy_field: bcs.bool()
} });
const InitCap = new MoveStruct({ name: `${$moduleName}::InitCap`, fields: {
  id: bcs.Address,
  publisher: _package.Publisher
} });
function initialize(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    null,
    null,
    null,
    null,
    null,
    "0x1::string::String",
    "0x1::string::String",
    "0x1::string::String",
    "0x1::string::String",
    "0x1::string::String",
    "0x1::string::String"
  ];
  const parameterNames = ["initCap", "advanceEpochApprover", "systemCurrentStatusInfo", "pricing", "supportedCurvesToSignatureAlgorithmsToHashSchemes", "dwalletCapImageUrl", "importedKeyDwalletCapImageUrl", "unverifiedPresignCapImageUrl", "verifiedPresignCapImageUrl", "unverifiedPartialUserSignatureCapImageUrl", "verifiedPartialUserSignatureCapImageUrl"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "ika_dwallet_2pc_mpc_init",
    function: "initialize",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  IKA_DWALLET_2PC_MPC_INIT,
  InitCap,
  initialize
};
